src/v/cluster/topics_frontend.cc

```diff
@@ -802,8 +809,12 @@ ss::future<topic_result> topics_frontend::do_create_topic(
       partition_allocator::shard,
       [assignable_config, topic_aware = _partition_autobalancing_topic_aware()](
         partition_allocator& al) {
+          if (assignable_config.has_custom_assignment()) {
+              return al.allocate(
+                make_custom_allocation_request(assignable_config, topic_aware));
```

Comment:
Mainly, we cannot be lazy about these paths, as the request has per-partition requirements.

Maybe we could improve this by delaying the transform, but I don't think there would be any obvious benefits.
The raw transformation takes an integer and instantiates an array of that size. 
This path translates an array into another array.
While it is very easy to request for a "billion" new partitions, it is not as easy to request for a billion custom assignments. If someone were to indeed ask for that, the point of failure would probably be around the parsing of the request and the population of this array. It would have to be handled there.